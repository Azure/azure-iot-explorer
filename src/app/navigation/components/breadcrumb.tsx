/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { NavLink } from 'react-router-dom';
import { BreadcrumbEntry as BreadcrumbEntryProps } from '../model';

export const Breadcrumb: React.FC<BreadcrumbEntryProps> = ({ disableLink, url, suffix, name }) => {
    return (
        <li className="breadcrumb-item">
            {disableLink && <>{name}</>}
            {!disableLink &&
                <NavLink to={`${url}${suffix || ''}`}>
                    {name}
                </NavLink>
            }
        </li>
    );
};
