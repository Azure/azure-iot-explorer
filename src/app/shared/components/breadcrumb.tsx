/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useSelector } from 'react-redux';
import { Route, NavLink, RouteComponentProps } from 'react-router-dom';
import { useLocalizationContext } from '../contexts/localizationContext';
import { ResourceKeys } from '../../../localization/resourceKeys';
import { ROUTE_PARTS, ROUTE_PARAMS } from '../../constants/routes';
import { getActiveAzureResourceHostNameSelector } from '../../azureResource/selectors';
import { getDeviceIdFromQueryString, getComponentNameFromQueryString } from '../utils/queryStringHelper';
import '../../css/_breadcrumb.scss';

const Breadcrumb = () => (
    <ul className="breadcrumb">
        <Route path="/:path" component={BreadcrumbItemContainer} />
    </ul>
);

export interface BreadcrumbItemDataProps extends RouteComponentProps{
    hostName: string;
}

export const BreadcrumbItem: React.FC<BreadcrumbItemDataProps> = props => {

    // tslint:disable-next-line:cyclomatic-complexity
    const renderBreadcrumbItem = (route: string | undefined) => {
        const { t } = useLocalizationContext();
        if (route) {
            switch (route) {
                case ROUTE_PARTS.RESOURCE: // redirects to login page
                    return <li className="breadcrumb-item"><NavLink to={'/'}>{'Hubs'}</NavLink></li>;
                case props.hostName:
                    return renderTextItem(getShortHubName());
                case ROUTE_PARTS.DEVICE_DETAIL:
                    return renderTextItem(getDeviceIdFromQueryString(props));
                case ROUTE_PARTS.DIGITAL_TWINS_DETAIL:
                    return renderTextItem(getComponentNameFromQueryString(props));
                case ROUTE_PARTS.MODULE_IDENTITY:
                        return renderLinkItem(getLocalizedKey(route), [ROUTE_PARAMS.DEVICE_ID]);
                case ROUTE_PARTS.DIGITAL_TWINS:
                    return renderLinkItem(getLocalizedKey(route), [ROUTE_PARAMS.DEVICE_ID]);
                default:
                    return renderLinkItem(getLocalizedKey(route));
            }
        }
        return <></>;
    };

    const getShortHubName = () => {
        return props.hostName && props.hostName.replace(/\..*/, '');
    };

    const getLocalizedKey = (route: string | undefined) => {
        const { t } = useLocalizationContext();
        const resourceKey = (ResourceKeys.breadcrumb as any)[route]; // tslint:disable-line:no-any
        return resourceKey ? t(resourceKey) : route;
    };

    const getCurrentUrl = () => {
        let url = props.match.url;
        if (url.endsWith('/')) {
            url = url.slice(0, -1); // remove trailing slash
        }
        return url.split('/').pop();
    };

    const renderLinkItem = (linkContent: string, searchParams?: ROUTE_PARAMS[]) => {
        const { match } = props;
        let href = match.url;
        if (searchParams) {
            href += '/?';
            searchParams.forEach((para, index) => {
                switch (para) {
                    case ROUTE_PARAMS.DEVICE_ID:
                        href += `${ROUTE_PARAMS.DEVICE_ID}=${getDeviceIdFromQueryString(props)}`;
                        break;
                    case ROUTE_PARAMS.COMPONENT_NAME:
                        href += `${ROUTE_PARAMS.COMPONENT_NAME}=${getComponentNameFromQueryString(props)}`;
                        break;
                    default:
                        break;
                }
                if (index !== searchParams.length - 1) {
                    href += '&';
                }
            });
        }
        return (
            <li className="breadcrumb-item">
                {match.isExact ?
                    linkContent :
                    <NavLink to={href}>
                        {linkContent}
                    </NavLink>
                }
            </li>
        );
    };

    const renderTextItem = (text: string) => {
        return (
            <li className="breadcrumb-item">
                {text}
            </li>
        );
    };

    return(
        <>
            {renderBreadcrumbItem(getCurrentUrl())}
            <Route path={`${props.match.url}/:path`} component={BreadcrumbItemContainer} />
        </>

    );
};

export type BreadcrumbItemContainerProps = RouteComponentProps;
export const BreadcrumbItemContainer: React.FC<BreadcrumbItemContainerProps> = props => {
    const viewProps = {
        hostName: useSelector(getActiveAzureResourceHostNameSelector),
        ...props
    };
    return <BreadcrumbItem {...viewProps} />;
};

export default Breadcrumb;
