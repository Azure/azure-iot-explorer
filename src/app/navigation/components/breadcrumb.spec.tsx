/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { shallow } from 'enzyme';
import { Breadcrumb } from './breadcrumb';

describe('Breadcrumb', () => {
    it('matches snapshot when link is disabled', () => {
        expect(shallow(
            <Breadcrumb
                disableLink={true}
                url={'url'}
                name={'name'}
                suffix={''}
                path={'path'}
            />
        )).toMatchSnapshot();
    });

    it('matches snapshot when link is enabled', () => {
        expect(shallow(
            <Breadcrumb
                disableLink={false}
                url={'url'}
                name={'name'}
                suffix={''}
                path={'path'}
            />
        )).toMatchSnapshot();
    });

    it('matches snapshot when suffix is truthy', () => {
        expect(shallow(
            <Breadcrumb
                disableLink={false}
                url={'url'}
                name={'name'}
                suffix={'suffix'}
                path={'path'}
            />
        )).toMatchSnapshot();
    });
});