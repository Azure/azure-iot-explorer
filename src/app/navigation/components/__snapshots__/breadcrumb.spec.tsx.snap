// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb matches snapshot when link is disabled 1`] = `
<li
  className="breadcrumb-item"
>
  name
</li>
`;

exports[`Breadcrumb matches snapshot when link is enabled 1`] = `
<li
  className="breadcrumb-item"
>
  <NavLink
    to="url"
  >
    name
  </NavLink>
</li>
`;

exports[`Breadcrumb matches snapshot when suffix is truthy 1`] = `
<li
  className="breadcrumb-item"
>
  <NavLink
    to="urlsuffix"
  >
    name
  </NavLink>
</li>
`;
